stan/math/prim/mat/vectorize/apply_vector_unary.hpp

```diff
@@ -0,0 +1,158 @@
+#ifndef STAN_MATH_PRIM_MAT_VECTORIZE_APPLY_VECTOR_UNARY_HPP
+#define STAN_MATH_PRIM_MAT_VECTORIZE_APPLY_VECTOR_UNARY_HPP
+
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/meta.hpp>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+// Forward declaration to allow specialisations
+template <typename T, typename Enable = void>
+struct apply_vector_unary {};
+
+/**
+ * Base template class for vectorization of unary vector functions
+ * defined by applying a functor to a standard library vector, Eigen dense
+ * matrix expression template, or container of these. For each specialisation,
+ * the same vector type as the input is returned.
+ *
+ * Two taxonomies of unary vector functions are implemented:
+ * - f(vector) -> vector
+ * - f(vector) -> scalar
+ *
+ * This base template class takes (and returns) Eigen expression templates.
+ */
+template <typename T>
+struct apply_vector_unary<T, require_eigen_t<T>> {
```

Comment:
Yep, this should also accept full matrices (which makes me think the name should be different, like ```apply_container_unary``` or something), since it allows for the easy extension of matrix functions to accept arrays of matrices. For example, this could be used to extend ```cholesky_decompose``` to also take ```std::vector<MatrixXd>``` inputs with minimal change to the code